### PR TITLE
Changing the Potion Event & Potion Recipe System to include more parameters for broader use

### PR DESCRIPTION
**Pull request by minzmann**
This PR adds a TileEntity parameter to the PotionBrewEvent, as well as to the getOutput method of IBrewingRecipe.

EDIT: now a World, a BlockPos and an Object parameter instead of TileEntity

Possible use cases:
- Limiting brewing of a potion to a dimension, biome or day time
- Spawning something near the brewing stand if a certain potion was brewed
- Giving nearby players a status effect when brewing is finished

Since the IRecipe change breaks compatibility, I updatet the PotionBrewEvent.Post to have the correct parent class (not the old deprecated one anymore) as well.

Also includes https://github.com/MinecraftForge/MinecraftForge/pull/2268 and the required changes to the test/example file.

Discussion/Feedback would be appreciated.

### CODE REVIEW

**Comment by liach** [comment]
Keep the non-tile entity sensitive version.

**Comment by minzmann** [comment]
This was also my first thought but without default methods in interfaces its imo just not worth it.
I thought of it this way: if your recipe doesn't care about the TileEntity, you should be able to pass null. Its just an extra burden if both methods are in the interface, then modders checking the recipes would have to call both methods to check, as well as people creating a recipe having to implement both methods. Same goes for the event, null could be passed.
Perhaps I should clarify this in the javadoc?

**Comment by liach** [comment]
I mean backward compatibility. Forge is based on Java 6 not Java 8 by the way.

**Comment by minzmann** [comment]
I know, thats why I said "without default interfaces". Due to them not existing, there is no way to add this method without breaking compatibility, since every Class that implements an interfaces has to have all its methods. And if compatibility is already broken, then there is no use in keeping the old one around.

File: src/main/java/net/minecraftforge/common/brewing/BrewingOreRecipe.java [hunk 1]
```diff
@@ -27,6 +27,7 @@ public boolean isIngredient(ItemStack stack)
             }
 
         }
+
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/brewing/BrewingOreRecipe.java]: Extra line

File: src/test/java/net/minecraftforge/test/BrewingRecipeRegistryTest.java [hunk 2]
```diff
@@ -21,19 +21,19 @@ public void init(FMLInitializationEvent event)
...
         BrewingRecipeRegistry.addRecipe(new ItemStack(Items.golden_sword), "gemDiamond", new ItemStack(Items.diamond_sword));
         
-        ItemStack output1 = BrewingRecipeRegistry.getOutput(new ItemStack(Items.golden_sword), new ItemStack(Items.diamond));
+        ItemStack output1 = BrewingRecipeRegistry.getOutput(new ItemStack(Items.golden_sword), new ItemStack(Items.diamond), null);
         if(output1.getItem() == Items.diamond_sword)
             System.out.println("Recipe succefully registered and working. Diamond Sword obtained.");
         
         // In vanilla, brewing netherwart into a water bottle results in an awkward potion (with metadata 16). The following tests if that still happens
-        ItemStack output2 = BrewingRecipeRegistry.getOutput(new ItemStack(Items.potionitem, 1, 0), new ItemStack(Items.nether_wart));
+        ItemStack output2 = BrewingRecipeRegistry.getOutput(new ItemStack(Items.potionitem, 1, 0), new ItemStack(Items.nether_wart), null);
```
Comment (liach) [inline src/test/java/net/minecraftforge/test/BrewingRecipeRegistryTest.java]: If you call with `null` as a `TileEntity` parameter, should you add `@Nullable` on the `TileEntity` parameter in the `getOutput` method in the interface to notify users that the tile entity might be null?

**Comment by liach** [comment]
Good. I like this pr. Will you add a hook for users to remove some existing brewing recipes, such as the non-tile sensitive vanilla brewing recipe?

**Comment by williewillus** [comment]
I think since we're moving major versions (people are starting to come from 1.7) it's okay to ditch the non-TE sensitive version

**Comment by minzmann** [comment]
@liach I'm not 100% sure on allowing to remove things, since getRecipes() was made to return an unmodifiable List, but i also think allowing people to remove other (or at least the vanilla ones) recipes would be a good things (e.g. a potion overhaul would be possible).

I'd like to hear @ganymedes01 opinion on this before changing it. (also this could be changed easily later down the line without breaking compatibility)

I'l take care of what you suggested soon, the annotation does make sense (allthough I don't see it being used a lot in forge), that extra line not so much.

**Comment by ganymedes01** [comment]
I think a better solution is passing the world and coordinates instead of the TIleEntity since that would be more flexible, other than that I think it's a good change.

**Comment by minzmann** [comment]
Passing coords and world would in fact be better, since it could have been triggered by an item or mob (not sure if someone is ever going to do this though, but if this thing is changed it doesn't hurt to include a broader use case). To signal this, im thinking of a boolean that indicates wether the recipe/event was triggered by a block/tileentity. It does seem a bit weird, so if you do have a better idea please tell me.

Something like this value is required, since the tile entity doesn't have to be a solid block (I'm pretty sure the vanilla one isn't), so checking for a block at coordinates might return a false positive if a player is standing inside the block bounds of a brewing stand.

**Comment by ganymedes01** [comment]
I think false positives would be really rare, but I suppose handling them wouldn't hurt.

Alternatively, you could simply add an ```Object``` parameter called crafter. It's then up to whatever is doing the crafting to pass itself through it and the recipes can handle it accordingly.

**Comment by Chimaine** [comment]
I'd prefer the `Object` parameter version. Otherwise, good change.

**Comment by minzmann** [comment]
Updated, now uses a World, a BlockPos and an Object parameter.

**Comment by minzmann** [comment]
Added methods for disabling vanilla recipes, as well as removing recipes.

**Comment by ganymedes01** [comment]
There's no need for the world and blockpos if you also have the crafter object :P
I think it should be either one or the other.

**Comment by minzmann** [comment]
In fact there is a need for both:
If only the Object is passed, whats supposed to be done with it? Even if you run a load of instanceof checks, while trying to find something that contains the location and world, you might end up with something that doesn't have these parameters, if you can even match the object with the instanceofs. This would just be a huge burden that wouldn't even guarantee a result.

And with only the location, we've got the scenario discussed earlier.

**Pull request closed without merging**